src/validator/LookupValidator.cpp

```diff
@@ -5,219 +5,162 @@
+    return DCHECK_NOTNULL(lookupCtx_)->schemaId;
 }
 
-Status LookupValidator::toPlan() {
-    auto* is = IndexScan::make(qctx_,
-                               nullptr,
-                               spaceId_,
-                               std::move(contexts_),
-                               std::move(returnCols_),
-                               isEdge_,
-                               schemaId_,
-                               isEmptyResultSet_);
-    is->setColNames(std::move(idxScanColNames_));
-    PlanNode* current = is;
-
-    if (withProject_) {
-        current = Project::make(qctx_, current, newYieldColumns_);
-    }
+GraphSpaceID LookupValidator::spaceId() const {
+    return DCHECK_NOTNULL(lookupCtx_)->space.id;
+}
 
-    if (dedup_) {
-        current = Dedup::make(qctx_, current);
+AstContext* LookupValidator::getAstContext() {
+    return lookupCtx_.get();
+}
 
-        // the framework will add data collect to collect the result
-        // if the result is required
-    }
+Status LookupValidator::validateImpl() {
+    lookupCtx_ = getContext<LookupContext>();
 
-    root_ = current;
-    tail_ = is;
+    NG_RETURN_IF_ERROR(prepareFrom());
+    NG_RETURN_IF_ERROR(prepareYield());
+    NG_RETURN_IF_ERROR(prepareFilter());
     return Status::OK();
 }
 
 Status LookupValidator::prepareFrom() {
-    auto* sentence = static_cast<const LookupSentence*>(sentence_);
-    spaceId_ = vctx_->whichSpace().id;
-    from_ = *sentence->from();
-    auto ret = qctx_->schemaMng()->getSchemaIDByName(spaceId_, from_);
-    if (!ret.ok()) {
-        return ret.status();
-    }
-    isEdge_ = ret.value().first;
-    schemaId_ = ret.value().second;
+    auto spaceId = lookupCtx_->space.id;
+    auto from = sentence()->from();
+    auto ret = qctx_->schemaMng()->getSchemaIDByName(spaceId, from);
+    NG_RETURN_IF_ERROR(ret);
+    lookupCtx_->isEdge = ret.value().first;
+    lookupCtx_->schemaId = ret.value().second;
     return Status::OK();
 }
 
 Status LookupValidator::prepareYield() {
-    auto* sentence = static_cast<const LookupSentence*>(sentence_);
-    returnCols_ = std::make_unique<std::vector<std::string>>();
-    // always return
-    if (isEdge_) {
-        returnCols_->emplace_back(kSrc);
-        idxScanColNames_.emplace_back(kSrcVID);
-        outputs_.emplace_back(idxScanColNames_.back(), vidType_);
-        returnCols_->emplace_back(kDst);
-        idxScanColNames_.emplace_back(kDstVID);
-        outputs_.emplace_back(idxScanColNames_.back(), vidType_);
-        returnCols_->emplace_back(kRank);
-        idxScanColNames_.emplace_back(kRanking);
-        outputs_.emplace_back(idxScanColNames_.back(), Value::Type::INT);
-    } else {
-        returnCols_->emplace_back(kVid);
-        idxScanColNames_.emplace_back(kVertexID);
-        outputs_.emplace_back(idxScanColNames_.back(), vidType_);
-    }
-    if (sentence->yieldClause() == nullptr) {
+    auto yieldClause = sentence()->yieldClause();
```

Comment:
It seems you forget to handle `outputs_`.

Reply:
fixed. thx.